cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -389,100 +397,183 @@ func (c ControllerContext) IsControllerEnabled(name string) bool {
 // that requests no additional features from the controller manager.
 // Any error returned will cause the controller process to `Fatal`
 // The bool indicates whether the controller was enabled.
-type InitFunc func(ctx context.Context, controllerCtx ControllerContext) (controller controller.Interface, enabled bool, err error)
+type InitFunc func(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller controller.Interface, enabled bool, err error)
+
+type RegistrableController interface {
```

Comment:
- can anybody come up with a better name?
- should I move this to a separate file?

Reply:
Yeah, I don't love the name.

`ControllerDescriptor` or something?  I am bad at naming things.

Reply:
Why do we need this as an interface?  For testing or will there be real implementations?

Reply:
I think simpler is best, so a struct with public members (no interface) plus well understood functions.  You could put this in a new file or even in sub-package.

Reply:
My main concern was to allow people to inject a custom logic into the function calls if needed. But I suppose, since we do not have no such use case yet, we could do without an interface.

It is hard to come up with a good name for this one.  Not sure about `Descriptor` yet, because it also contains the logic for running the controllers... 

Also not sure how to name the functions: for now I have `newEndpointsRegistrableController` for the following reasons:
1. should not refer to creating or starting the controller, since that is done elsewhere
2. should not have a long prefix so the original controller name is easily readable, so suffix is preferable IMO

Reply:
This is not a public API (at least not outside of k/k) so we can always change it in the event we really do need an interface.  But it seems like a "dumb data object" is appropriate.  Even with a function pointer.

I like `Descriptor` or `Handle` because this object is the beginning of a controller.  It's not running yet.  It holds metadata, but we have too many `*Meta` things.

Alternately, you could make this all part of `controller.Interface` and document that lifecycle as having a pre-running phase.  That spreads out the metadata more and puts it near the controller implementation, which I am not sure is right.  This feels like an "inode vs. file" thing - many inodes (names and aliases) can refer to the same file (descriptor).  That said, the only method in `controller.Interface` is `Name()` and it seems like many of these `start` functions return `nil` (e.g. `startServiceController`) for the controller.Interface anyway!!  That should be a cleanup of its own.

`newEndpointsControllerDescriptor` is long but not terrible.  It says what it means.  `newEndpointsControllerDesc` is a bit shorter and not much less readable.  Either one satisfies your criteria.  

So I would say either something like

```
func newFooControllerDescriptor() ControllerDescriptor {
    return ControllerDescriptor{
        Name: "foo",
        Aliases: []string{"foo-ctrl"},
        Run: startFooController,
    }
}

func startFooController() error {
    // lots of stuff
}
```

or else bake this into `controller.Interface` like:

```
type Interface interface {
    Name() string
    Aliases() []string
    Run(ctrlCtxControllerContext ) error
}
````

and then just call `NewFooController() controller.Interface`, check gates, then call the Run() method on it.

Reply:
I suppose we can look at it as a descriptor - I like the suggestions. I would prefer we leave `controller.Interface` refactoring out of this PR, and I am not yet convinced that we should move the functionality there. Nevertheless, there should not be too many changes in the code that this PR touches, if we decide to move it.

Reply:
renamed to ControllerDescriptor